### DISCUSSION TITLE
Need Some Help in Message_Passing Class

### DISCUSSION THREAD

**Discussion by utkarsh0902311047** [pyg-team/pytorch_geometric#8227] [Q&A]
In the message passing class, there is a propagate function. I want to collect all the results of "_collect," especially pos[i] , pos[j], edge_index in a file every time it goes through it doesn't matter how many epochs or how many times it goes through _collect in an epoch. I want to collect all these values in a file in readable format.
How can I do it?

**Comment by rusty1s** [comment]
You can leverage hooks for this, see https://pytorch-geometric.readthedocs.io/en/latest/generated/torch_geometric.nn.conv.MessagePassing.html#torch_geometric.nn.conv.MessagePassing.register_propagate_forward_pre_hook.

**Reply by utkarsh0902311047** [reply]
Sorry, I have never used hooks before. Please can you give me a little more insight on how to use them to achieve my above-mentioned requirement?

**Reply by rusty1s** [reply]
After instantiating your GNN layer, you can register a hook to it
```python
def hook(module, inputs):
    # parse inputs and print/store them.

conv.register_propagate_forward_pre_hook(hook)
```
`hook` will be called before each execution of `propagate`.

**Reply by utkarsh0902311047** [reply]
In pointnet2_classification.py, I write a function hook:
```
def hook(module, input):
       print(f"Forward pre-hook called for module: {module.__class__.__name__}")
       print(f"Input data: {input}")
```
Then after ```model = Net().to(device)```, I write 
```PointNetConv.register_propagate_forward_pre_hook(hook)```. 
But it gives error:
```TypeError: register_propagate_forward_pre_hook() missing 1 required positional argument: 'hook'```

1. What am I doing wrong?
2. How to save all "_collect" of message passing in a file for all layers and epoch?

**Reply by rusty1s** [reply]
You need to register on the instantiated object, not the class. So it would be something like `self.conv.register_propagate_forward_pre_hook(hook)`.

**Reply by utkarsh0902311047** [reply]
```self.conv.register_propagate_forward_pre_hook(hook)
NameError: name 'self' is not defined
```
I am getting this error

**Reply by rusty1s** [reply]
Depends on where your layers are defined. How does your model look like?

**Reply by utkarsh0902311047** [reply]
The model is pointnet2_classification.py

**Reply by rusty1s** [reply]
Then this would be
```python
model.sa1_module.conv.register_propagate_forward_pre_hook(...)
